### DISCUSSION TITLE
Individual hd EEG positions

### DISCUSSION THREAD

**Discussion by idossha** [simnibs/simnibs#302] [7.Creating head models]
Hi SimNIBS team,

I wanted to see if there is any streamlined way of doing the following:

I have individual MRI scans, and digitization of hd-EEG locations for respective subjects.

1. What will be the steps to fit the EEG cap to the .msh?
      - Is there a way to automate it such that the charm function will do it?
2. What will be the best approach to create a field summary output for TI.py?

Thanks,
Ido

**Comment by jduemose** [comment]
Hi Ido

Let me try to answer (1):

This is not really part of the standard simnibs pipeline, however, if you have digitized electrode positions as well as fiducial landmarks (e.g., LPA, RPA, and nasion), then you can locate those on the MRI of the subject, and find the transformation that takes the electrodes from "electrode space" to MRI space (which is the same as that of the msh file).
Although not officially documented, the Montage class found here https://github.com/simnibs/simnibs/blob/master/simnibs/eeg/utils.py should be able to do this for you: make one that contains the electrodes positions, names, and landmarks (all in digitized space), and another containing the fiducials in MRI space. Then use the `fit_to` method to fit the former to the latter.
Have a look and let me know if you need help.

Best,
Jesper

**Reply by idossha** [reply]
Hi Jesper,

Thank you so much for your response. I was able to figure out a way to create field summary for my TI. But I was not able to follow your instructions successfully. 

I am really unclear to how I should approach this... I tried multiple approaches without much progress. 

Could you give me higher resolution instructions?

---

Worth mentioning:

I do individual .csv of hd-EEG net (EGI) for my participants. For a sanity check, I did try to look just at the fiducial based on the MRI scan as you suggested, and it looks great. So the only missing piece is the transformation of the rest of the net based on the fiducial. 

Again, thanks for your help with this.

Ido

**Comment by jduemose** [comment]
Hi Ido

I apologize for the late response but I was on vacation!

Let us assume that you have your digitized electrodes positions in a csv file formatted in a way that simnibs can read (e.g., look at the warped templates in the charm run folder of a subject)
```
Electrode,-27.5,81.7,-5.2,Fp1
Electrode,30,81.7,-5.4,Fp2
Electrode,0,58,66,Fz
...
```
saved as `digitized_positions.csv` and the corresponding fiducials/landmarks like
```
Fiducial,0,82.9,-43,Nz
Fiducial,-79.6,-20.6,-48.6,LPA
Fiducial,80.6,-20.6,-48.1,RPA
```
saved as `eeg_fiducials.csv`[^1]. Now, put the MRI fiducial positions in a similarly formatted file; let us call this `mri_fiducials.csv`. Then, try the following
```
montage = simnibs.eeg.utils.Montage.from_filename("digitized_positions.csv")
montage.add_landmarks("eeg_fiducials.csv")

fid_mri = simnibs.eeg.utils.Montage()
fid_mri.add_landmarks("mri_fiducials.csv")

trans = montage.fit_to(fid_mri) # affine trans that maps matched points (eeg fids to mri fids) in a least sq. sense
montage.apply_trans(trans) # now in mri space

montage.write("digitized_positions_mri_space.csv")
```
Let me know if this works or not :)

[^1]: Normally you should be able to store electrodes and fiducials in the same file, however, I discovered a bug in the code constructing a `Montage` from such a file while trying it out so it is probably easiest to keep it in separate files for now. I will fix it in a future release!).

**Reply by idossha** [reply]
Awesome. No worries about the delay, I totally appreciate the time you are taking to help. I will try it over the weekend and let you know in a few days. Thanks again!

**Reply by idossha** [reply]
Hi Jesper,

Sorry for the long delay. I tried your suggestion and it did not work for me. Currently I found a workaround where I use the transformation of another toolbox, and transform the output to a format that is compatible with simnibs. So I am good to work with individual caps for now. Thanks for your help!

Ido
